Work tracking example for external data sync

ChumpChief opened the pull request:
This PR adds a stubbed example that we'll iterate on to demonstrate sync'ing with an external data source.  It intends to simulate sync'ing with an external work tracking service to explore this area.

Most interesting files are:

- taskList.ts -- I expect this is where much of the sync'ing logic will live (i.e. it is the right place to encapsulate knowledge of "what is a task list, where does it come from, what's its data look like").  Lots to be done here.
- externalData.ts -- Very simple approach to simulate using an external data source.  We may eventually need to replace this with something more robust (i.e. a real Express server to talk to) but should be sufficient for initial investigations at least.
- The views -- I think we need to demonstrate two things -- (1) a reasonable user experience with our proposals and (2) sufficient insight into how things work that our customers can understand and be successful with their own implementations.  The views we present for the demo are a big part of achieving these goals.

Opening as a draft as I may want to stub out a little bit more before the first checkin, but what's there should be relatively interesting already.

ChumpChief committed:
Add work-tracking example

ChumpChief committed:
One model

ChumpChief committed:
View cleanup

ChumpChief committed:
Remove migrator stuff

ChumpChief committed:
Remove migrator

ChumpChief committed:
Remove migration tool

ChumpChief committed:
Remove more interfaces

ChumpChief committed:
Some renaming

ChumpChief committed:
More rename

ChumpChief committed:
Start prep for data import

ChumpChief committed:
Initial export

ChumpChief committed:
Move initial data import into the data object

ChumpChief committed:
Remove migrationInterfaces

ChumpChief committed:
Friendlier names

ChumpChief committed:
Add TODOs

ChumpChief committed:
Rename to external-data

ChumpChief committed:
Update readme/package/test

ChumpChief committed:
Use StaticCodeLoader

ChumpChief committed:
Move parsing into externalData.ts

ChumpChief committed:
Use TinyliciousModelLoader

ChumpChief committed:
Update some comments, scope some exposure

ChumpChief committed:
More scoping

ChumpChief committed:
Use localStorage and add controls

ChumpChief committed:
Use a table for view

ChumpChief committed:
Add external data view

ChumpChief committed:
Update version

ChumpChief committed:
Update some comments, switch to TypedEventEmitter

ChumpChief committed:
Bunch more TODOs

ChumpChief committed:
One more TODO before draft PR

ChumpChief committed:
Merge remote-tracking branch 'upstream/main' into WorkTrackingExample

ChumpChief committed:
Update package versions

WayneFerrao commented on examples/hosts/app-integration/external-data/src/model/taskList.ts:
```diff
@@ -0,0 +1,190 @@
+        nameString.insertText(0, name);
+        const priorityCell: SharedCell<number> = SharedCell.create(this.runtime);
+        priorityCell.set(priority);
+        this.root.set(id, { id, name: nameString.handle, priority: priorityCell.handle });
```
Shouldn't this be this.tasks.set?

ChumpChief replied:
Good question, I'll add some comments in the code to clarify too --

Here I set the value on this.root to make the actual data change (and so other clients will see the new member of the root).  The this.tasks collection really just holds these local facades (Task class) on the data for individual items for convenience and encapsulation (flushes out the async so we can use a sync interface, give a convenient object to hand to a view rather than the whole this.root SharedMap).  So changing the this.tasks collection alone wouldn't result in any persisted data changes.

Instead, when this.root has a "valueChanged" event (which happens immediately when setting the new content on the root) we'll create and add the Task to the this.tasks collection (in handleTaskAdded).  Changing this.root triggers the change in this.tasks automatically.

Overall it's a slight variation on the collection pattern (see this demo for a more thorough dive on that pattern: https://github.com/microsoft/FluidFramework/tree/main/examples/apps/contact-collection).

Josmithr commented:
If we're adding a new package, we should make sure it's set up with prettier, since there is a parallel effort to get it running in all existing packages.

ChumpChief commented:
> If we're adding a new package, we should make sure it's set up with prettier, since there is a parallel effort to get it running in all existing packages.

Sounds good - do you have a pointer to what I need to do for that?

ChumpChief committed:
Add comments on tasks collection

ChumpChief committed:
Add deleteTask

ChumpChief committed:
Remove disabled state unless/until needed

ChumpChief committed:
Add some docs, more TODO, readonly view props

ChumpChief committed:
More docs, more TODO, change debug event name

ChumpChief commented on examples/hosts/app-integration/external-data/src/externalData.ts:
```diff
@@ -0,0 +1,78 @@
+/**
+ * Parse string data into an array of simple objects that are easily imported into an
+ * task list.
+ * @param stringData - formatted string data
```
Made a quick note here, but also a TODO that we might want to move away from the string format here.

ChumpChief committed:
Add an inheritDoc for the container runtime factory

ChumpChief commented on examples/hosts/app-integration/external-data/src/externalData.ts:
```diff
@@ -0,0 +1,78 @@
+
+const localStorageKey = "fake-external-data";
+
+export interface IExternalDataSourceEvents extends IEvent {
```
Added docs here and a number of other places

ChumpChief committed:
Add prettier

ChumpChief committed:
Update other users of prettier to 2.6.2

ChumpChief committed:
Update package lock

ChumpChief committed:
Add reference to ignore file

ChumpChief committed:
Apply some prettier fixes

ChumpChief commented:
Added the prettier scripts and updated the dependency, also applied some of its suggestions.  I will say though that the formatting it suggested was frequently pretty ugly, so I'm a little worried about what might happen if we turn prettier on generally.

Josmithr commented:
> Added the prettier scripts and updated the dependency, also applied some of its suggestions. I will say though that the formatting it suggested was frequently pretty ugly, so I'm a little worried about what might happen if we turn prettier on generally.

There are definitely some formatting choices prettier makes that I don't find particularly appealing, but the general idea is to let prettier be opinionated in favor of consistency across our code (and not have to have discussions about formatting in PR reviews and such).

That said, we do have a shared config that sets a few rules where prettier gives options. We probably want to add a prettier config file to the directory and have it extend our base config in `build-common`. That might result in some positive changes.

ghoshkaj commented on examples/hosts/app-integration/external-data/README.md:
```diff
@@ -0,0 +1,49 @@
+    - For an even faster build, you can add the package name to the build command, like this:
+      `npm run build:fast -- --nolint @fluid-example/app-integration-schema-upgrade`
+1. In a separate terminal, start a Tinylicious server by following the instructions in [Tinylicious](../../../server/tinylicious).
+1. Run `npm run start` from this directory (examples/hosts/app-integration/schema-upgrade) and open <http://localhost:8080> in a web browser to see the app running.
```
```suggestion
1. Run `npm run start` from this directory (examples/hosts/app-integration/external-data) and open <http://localhost:8080> in a web browser to see the app running.
```

ghoshkaj commented on examples/hosts/app-integration/external-data/README.md:
```diff
@@ -0,0 +1,49 @@
+
+1. Run `npm install` and `npm run build:fast -- --nolint` from the `FluidFramework` root directory.
+    - For an even faster build, you can add the package name to the build command, like this:
+      `npm run build:fast -- --nolint @fluid-example/app-integration-schema-upgrade`
```
```suggestion
      `npm run build:fast -- --nolint @fluid-example/app-integration-external-data`
```

ChumpChief replied:
Uh oh my copy/paste is showing :-P

ChumpChief committed:
Fix schema-upgrade -> external-data

ChumpChief committed:
More comments and todo details

ChumpChief commented on examples/hosts/app-integration/external-data/src/model/taskList.ts:
```diff
@@ -0,0 +1,238 @@
+        // and persisted.  In turn, this will trigger the "valueChanged" event and handleTaskAdded which will update
+        // the this.tasks collection.
+        this.root.set(id, { id, name: nameString.handle, priority: priorityCell.handle });
+        // TODO: Ultimately we want to retain the data we retrieved from the external source separate from the draft
```
@WayneFerrao I think this will be the best TODO to start on -- maybe we can chat about this a bit tomorrow too?

ghoshkaj commented on examples/hosts/app-integration/external-data/README.md:
```diff
@@ -0,0 +1,49 @@
+1. Run `npm install` and `npm run build:fast -- --nolint` from the `FluidFramework` root directory.
+    - For an even faster build, you can add the package name to the build command, like this:
+      `npm run build:fast -- --nolint @fluid-example/app-integration-external-data`
+1. In a separate terminal, start a Tinylicious server by following the instructions in [Tinylicious](../../../server/tinylicious).
```
```suggestion
2. In a separate terminal, start a Tinylicious server by following the instructions in [Tinylicious](../../../server/tinylicious).
```

ChumpChief replied:
This is actually just a little markdown trick - if you use 1. for everything, it will still render with correct numbers.  Makes it a little more convenient if later you add more items to the middle or rearrange them, so you don't have to manually renumber.

ghoshkaj commented on examples/hosts/app-integration/external-data/README.md:
```diff
@@ -0,0 +1,49 @@
+    - For an even faster build, you can add the package name to the build command, like this:
+      `npm run build:fast -- --nolint @fluid-example/app-integration-external-data`
+1. In a separate terminal, start a Tinylicious server by following the instructions in [Tinylicious](../../../server/tinylicious).
+1. Run `npm run start` from this directory (examples/hosts/app-integration/external-data) and open <http://localhost:8080> in a web browser to see the app running.
```
```suggestion
3. Run `npm run start` from this directory (examples/hosts/app-integration/external-data) and open <http://localhost:8080> in a web browser to see the app running.
```
